File: 230. Kth Smallest Element in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    let stack = [root];
    while (stack.length > 0)
    {
        while (root && root.left)
        {
            stack.push(root.left);
            root = root.left;
        }
        root = stack.pop();
        k -= 1;
        if (k === 0)
            return root.val;
        if (root.right)
            stack.push(root.right);
        root = root.right;
    }
};

File: 230. Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack <TreeNode*> s;
        s.push(root);
        while (!s.empty())
        {
            while (root != NULL && root->left != NULL)
            {
                s.push(root->left);
                root = root->left;
            }
            root = s.top();
            s.pop();
            if (--k == 0)
                return root->val;
            if (root->right)
                s.push(root->right);
            root = root->right;
        }
        return 0;
    }
};
